domain/storageprovisioning/service/service.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/juju/core/changestream"
+	"github.com/juju/juju/core/machine"
+	"github.com/juju/juju/core/watcher"
+	"github.com/juju/juju/core/watcher/eventsource"
+	machineerrors "github.com/juju/juju/domain/machine/errors"
+	"github.com/juju/juju/internal/errors"
+)
+
+// Service provides the interface for provisioning storage within a model.
+type Service struct {
+	watcherFactory WatcherFactory
+
+	st State
+}
+
+// State is the accumulation of all the requirements [Service] has for
+// persisting and watching changes to storage instances in the model.
+type State interface {
+	FilesystemState
+	VolumeState
+
+	// CheckMachineIsDead checks to see if a machine is dead returning true when
+	// the life of the machine is dead.
+	//
+	// The following errors may be returned:
+	// - [github.com/juju/juju/domain/machine/errors.MachineNotFound] when no
+	// machine exists for the provided uuid.
+	CheckMachineIsDead(context.Context, machine.UUID) (bool, error)
+
+	// GetMachineNetNodeUUID retrieves the net node uuid associated with provided
+	// machine.
+	//
+	// The following errors may be returned:
+	// - [github.com/juju/juju/domain/machine/errors.MachineNotFound] when no
+	// machine exists for the provided uuid.
+	GetMachineNetNodeUUID(context.Context, machine.UUID) (string, error)
+
+	// NamespaceForWatchMachineCloudInstance returns the change stream namespace
+	// for watching machine cloud instance changes.
+	NamespaceForWatchMachineCloudInstance() string
+}
+
+// WatcherFactory instances return watchers for a given namespace and UUID.
+type WatcherFactory interface {
+	// NewNamespaceMapperWatcher returns a new watcher that receives changes
```

Comment:
I wish we didn't do this. This will now be the 8th place that copies this block of text and which will become out of date if we amend the watcher factory doc comment.